script.js

```diff
@@ -0,0 +1,167 @@
+const countriesGrid = document.querySelector('.countries-grid');
+let countryFetchedData = null;
+
+async function fetchCountriesData() {
+    if (countryFetchedData) {
+        return;
+    }
+    try {
+        const response = await fetch('./CountriesData.json');
+        if (!response.ok) {
+            throw `Failed to fetch countries data with status - ${response.status}`;
+        }
+        countryFetchedData = await response.json();
+    } catch (error) {
+        console.error('Error fetching countries data:', error);
+    }
+}
+
+function getDropdownsRegions(data) {
+    const dropdownRegionFilter = document.querySelector('.dropdown-body');
+    const uniqueRegionsSet = new Set(data.map(country => country.region));
+    const uniqueRegions = [...uniqueRegionsSet]
+
+    const dropdownList = document.createElement('ul');
+    const allOption = document.createElement('li');
+    allOption.setAttribute('data-region', 'all');
+    allOption.textContent = 'All';
+    dropdownList.appendChild(allOption);
+
+    uniqueRegions.forEach(region => {
+        const dropdownItem = document.createElement('li');
+        dropdownItem.setAttribute('data-region', region);
+        dropdownItem.textContent = region;
+        dropdownList.appendChild(dropdownItem);
+    });
+    dropdownRegionFilter.appendChild(dropdownList);
+}
+
+function getSearchdata() {
+    const searchInput = document.querySelector('.search-input');
+    searchInput.addEventListener('input', (e) => {
+        searchValue = e.target.value;
+        const filteredData = filterDataSearch(countryFetchedData, searchValue);
+        countriesGrid.innerHTML = '';
+        analyzeMainContainer(filteredData);
+    });
+}
+
+function filterDataSearch(data, searchValue) {
+    return data.filter(({ name }) => name.toLowerCase().includes(searchValue.toLowerCase()));
+}
+
+function analyzeMainContainer(data) {
+    data.forEach(({ name, flag, population, region, capital }) => {
+        const countryCard = document.createElement('a');
+        countryCard.href = './details.html';
+        countryCard.addEventListener('click', function () {
+            const countryData = { name, flag, population, region, capital };
+            localStorage.setItem('countryDetails', JSON.stringify(countryData));
+        });
+
```

Comment:
Instead of using localStorage like this, create a service that expose method related to localStorage. 

This will help you avoid repeat your code and seperate different concepts